users/user-guide-query/multi-stage-query/operator-types/leaf.md

```diff
@@ -0,0 +1,238 @@
+
+This along with `numEntriesScannedInFilter` indicates where most of the time is spent during table scan processing.
+A high number for this means the selectivity is low (that is, Pinot needs to scan a lot of records to answer the query).
+If this is high, consider using star-tree index, given a regular index won't improve performance.
+
+### numSegmentsQueried
+Type: Integer
+
+Similar to the same stat in single-stage queries, this stat indicates the total number of segment queried for a query. 
+May be less than the total number of segments if the broker applies optimizations. <!-- Can this happen in V2? -->
+
+The broker decides how many segments to query on each server, based on broker pruning logic.
+The server decides how many of these segments to actually look at, based on server pruning logic. 
+After processing segments for a query, fewer may have the matching records. 
+
+In general, `numSegmentsQueried >= numSegmentsProcessed >= numSegmentsMatched`.
+
+### numSegmentsProcessed
+Type: Integer
+
+Similar to the same stat in single-stage queries, this stat indicates the number of segments processed with at least one
+document matched in the query response.
+
+The more segments are processed, the more IO has to be done.
+This is why selective queries where `numSegmentsProcessed` is close to `numSegmentsQueried` can be optimized by
+changing the data distribution.
+
+### numSegmentsMatched
+Type: Integer
+
+Similar to the same stat in single-stage queries, this stat indicates the number of segment operators used to process 
+segments. Indicates the effectiveness of the pruning logic.
+
+### totalDocs
+Type: Long
+
+Similar to the same stat in single-stage queries, this stat indicates the number of rows in the table.
+
+### numGroupsLimitReached
+Type: Boolean
+
+Similar to the same stat in single-stage queries and the same in [aggregate](./aggregate.md#numgroupslimitreached) 
+operators, this stat indicates if the max group limit has been reached in a `group by` aggregation operator executed
+in the leaf stage. 
+
+If this boolean is set to true, the query result may not be accurate. 
+The default value for `numGroupsLimit` is 100k, and should be sufficient for most use cases.
+
+### numResizes
+Type: Integer
+
+Number of result resizes for queries
+
+### resizeTimeMs
+Type: Long
+
+Time spent in resizing results for the output. 
+Either because of LIMIT or maximum allowed group by keys or any other criteria.
+
+### threadCpuTimeNs
+Type: Long
+
+Aggregated thread cpu time in nanoseconds for query processing from servers.
+This metric is only available if Pinot is configured with `pinot.server.instance.enableThreadCpuTimeMeasurement`.
+
+### systemActivitiesCpuTimeNs
+Type: Long
+
+Aggregated system activities cpu time in nanoseconds for query processing (e.g. GC, OS paging etc.)
+This metric is only available if Pinot is configured with `pinot.server.instance.enableThreadCpuTimeMeasurement`.
+
+### numSegmentsPrunedByServer
```

Comment:
how to think about these numSegmentsPruned* vs numSegmentsQueried / Processed / Matched? Are these at the operator level vs the latter on a query level?
also, i'm confused why these are on a leaf operator level, there can be more than 1 leaf operator on a server right?

Reply:
> how to think about these numSegmentsPruned* vs numSegmentsQueried / Processed / Matched? Are these at the operator level vs the latter on a query level?

Yes, this is operator level and the others are the query level, which are basically the sum of the operator level.

> also, i'm confused why these are on a leaf operator level, there can be more than 1 leaf operator on a server right?

Yes, there can be more than 1 leaf operator on a server. Specifically, imagine we have a query joining two tables. Each table scan will scan a table using different filters. Each table scan will be one leaf operator. Given each one may have different filters (and touch different tables) each will probably prune a different number of segments.

Now imagine that the number of segments pruned is very low at query level. You can then analyze the pruning of each leaf operator and see which is the table that is not being pruned as expected and therefore try to focus that table knowing that the other table is scanned efficiently.